src/libutil/logging.hh

```diff
@@ -225,7 +225,11 @@ inline void warn(const std::string & fs, const Args & ... args)
     logger->warn(f.str());
 }
 
-void warnOnce(bool & haveWarned, const FormatOrString & fs);
+#define warnOnce(haveWarned, args...) \
+    if (!haveWarned) {                \
+        haveWarned = true;            \
+        warn(args);                   \
+    }
```

Comment:
What's the point of using a macro here?

Reply:
To ensure that the arguments are evaluated lazily, just like `debug()` etc.